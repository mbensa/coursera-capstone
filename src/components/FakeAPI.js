const seededGenerator = (date, hour) => {
  const parsedDate = new Date(date); // Parse the date string into a Date object
  const m = 9;
  const d = parsedDate.getDate(); // Use getDate() method on the parsed date object
  const result = ((d + hour) % m) / 10;

  return result;
};

const fetchAPI = (date) => {
  let result = [];

  result.push("--- Select a Time ---");

  for (let hour = 15; hour <= 23; hour++) {
    if (seededGenerator(date, hour) < 0.5) result.push(hour + ":00");
    if (seededGenerator(date, hour + 7) < 0.5) result.push(hour + ":30");
  }

  return result;
};

const submitAPI = (formData) => true;

const fakeAPI = {
  fetchAPI: fetchAPI,
  submitAPI: submitAPI,
};

export default fakeAPI;
