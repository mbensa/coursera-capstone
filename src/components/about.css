.aboutImgContainer {
    height: 200px;
    background-image: url(../assets/marioAndadrian.jpg);
    background-size: cover;
}

.aboutTextContainer {
    padding: 20px;
}

@media (min-width: 1024px) {
    .aboutContainer {
        display: flex;
        padding: 20px;
        justify-content: center;
    }

    .aboutTextContainer {
        padding: 0 20px;
    }

    .aboutImgContainer {
        width: 350px;
    }
}