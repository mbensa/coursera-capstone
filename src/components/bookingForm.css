.formContainer {
    font-family: 'Karla', sans-serif;
    gap: 20px;
    display: flex;
    flex-direction: column;
}

.submitButton {
    padding: 19px 24px;
    border: none;
    cursor: pointer;
    border-radius: 8px;
    background-color: var(--secondary);
    width: 200px;
    display: flex;
    justify-content: center;
  }

  input {
    font-family: 'Karla', sans-serif;
    padding: 10px;
}

select {
    padding: 10px;
}

  @media (max-width: 1024px) {
    .submitButton {
      padding: 8px 16px;
      border-radius: 16px;
    }
  }

  @media (min-width: 1024px) {
    .formContainer {
        max-width: 500px;
        margin: 0 auto;
    }

    .submitButton {
        margin: 50px auto;
    }
  }