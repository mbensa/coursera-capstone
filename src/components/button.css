/********************************
********Button Component*********
********************************/
.btn {
  padding: 19px 24px;
  border: none;
  cursor: pointer;
  border-radius: 8px;
  background-color: var(--secondary);
  width: 200px;
  display: flex;
  justify-content: center;
}


@media (max-width: 1024px) {
  .btn {
    padding: 8px 16px;
    border-radius: 16px;
    width: 150px;
  }
}