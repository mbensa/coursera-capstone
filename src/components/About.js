import React from "react";
import Text from "./Text";
import "./about.css";
import useMobile from "../hooks/useMobile";

export default function About() {
  const { isMobile } = useMobile();

  if (isMobile) {
    return (
      <section>
        <div className="aboutImgContainer"></div>
        <div className="aboutTextContainer">
          <Text type="h1">Little Lemon</Text>
          <Text type="h2">Chicago</Text>
          <Text type="h5">
            Little Lemon is run by two famous chefs, Mario and Adrian. The
            restaurant has opened its doors in 2003. We serve Mediterranean food
            with a modern twist.{" "}
          </Text>
        </div>
      </section>
    );
  } else {
    return (
      <section className="aboutContainer">
        <div className="aboutTextContainer">
          <Text type="h1">Little Lemon</Text>
          <Text type="h2">Chicago</Text>
          <Text type="p">
            Little Lemon is run by two famous chefs, Mario and Adrian. The
            restaurant has opened its doors in 2003. We serve Mediterranean food
            with a modern twist.{" "}
          </Text>
        </div>
        <div className="aboutImgContainer"></div>
      </section>
    );
  }
}
