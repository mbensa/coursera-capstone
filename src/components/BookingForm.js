import React, { useState, useEffect } from "react";
import Nav from "./Nav";
import Text from "./Text";
import Footer from "./Footer";
import "./bookingForm.css";

function BookingForm({
  date,
  setDate,
  name,
  setName,
  time,
  setTime,
  guests,
  setGuests,
  occasion,
  setOccasion,
  availableTimes,
  updateTimes,
  submitForm, // Receive submitForm as a prop
}) {
  const [formData, setFormData] = useState({
    date: "",
    time: "",
    guests: 1,
    occasion: "Birthday",
    name: "",
  });

  const [formValid, setFormValid] = useState(false);

  useEffect(() => {
    // Check if all form fields are valid
    const isValid = name && date && time && guests >= 1 && occasion !== "";

    // Update form validity state
    setFormValid(isValid);
  }, [date, time, guests, occasion, name]);

  const handleDateChange = (event) => {
    const newDate = event.target.value;
    setFormData((prevData) => ({
      ...prevData,
      date: newDate,
    }));
    updateTimes(newDate);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (formValid) {
      submitForm({ date, time, guests, occasion, name });
    }
  };

  return (
    <>
      <Nav />
      <main className="mainContainerBooking">
        <Text type="h1">Book a table</Text>
        <div className="restaurantImg"></div>
        <form className="formContainer" onSubmit={handleSubmit}>
          <label htmlFor="res-name">
            <Text type="h5">Name and Surname</Text>
          </label>
          <input
            type="text"
            id="res-name"
            value={name}
            onChange={(e) => setName(e.target.value)}
            required
            aria-label="Name input"
          />
          <label htmlFor="res-date">
            <Text type="h5">Choose Date</Text>
          </label>
          <input
            data-testid="resDate"
            type="date"
            id="res-date"
            value={date}
            onChange={handleDateChange}
            min={new Date().toISOString().split("T")[0]}
            required
            aria-label="Date Selection"
          />
          <label htmlFor="res-time">
            <Text type="h5">Choose Time</Text>
          </label>
          <select
            id="res-time"
            value={time}
            onChange={(e) => setTime(e.target.value)}
            required
            aria-label="Time Selection"
          >
            {availableTimes.map((availableTime, index) => (
              <option key={index}>{availableTime}</option>
            ))}
          </select>
          <label htmlFor="guests">
            <Text type="h5">Number of Guests</Text>
          </label>
          <input
            type="number"
            placeholder="1"
            min="1"
            max="10"
            id="guests"
            value={guests}
            onChange={(e) => setGuests(parseInt(e.target.value))}
            required
            aria-label="Guests Selection"
          />
          <label htmlFor="occasion">
            <Text type="h5">Occassion</Text>
          </label>
          <select
            id="occasion"
            value={occasion}
            onChange={(e) => setOccasion(e.target.value)}
            required
            aria-label="Occasion Selection"
          >
            <option>Birthday</option>
            <option>Anniversary</option>
          </select>
          <input
            className="submitBtn"
            type="submit"
            value="Make Your reservation"
            disabled={!formValid}
            aria-label="Submit Reservation"
          />
        </form>
      </main>
      <Footer />
    </>
  );
}

export default BookingForm;
