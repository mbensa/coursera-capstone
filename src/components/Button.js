import React from "react";
import Text from "./Text";
import "./button.css";

export default function Button(props) {
  const { children } = props;

  return (
    <div className={"btn"}>
      <Text type="h5">{children}</Text>
    </div>
  );
}
