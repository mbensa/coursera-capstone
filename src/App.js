import React, { useState, useReducer, useEffect } from "react";
import "./App.css";
import BookingForm from "./components/BookingForm";
import MainPage from "./pages/MainPage";
import ConfirmedBooking from "./components/ConfirmedBooking"; // Import the ConfirmedBooking component
import fakeAPI from "./components/FakeAPI";
import { Route, Routes, useNavigate } from "react-router-dom";

// Define reducer function to handle state changes
const updateTimesReducer = (state, action) => {
  switch (action.type) {
    case "INITIALIZE_TIMES":
      return initializeTimes(); // Initialize availableTimes
    case "UPDATE_TIMES":
      return updateTimes(state, action.payload); // Update availableTimes based on selected date
    default:
      return state;
  }
};

export const initializeTimes = () => {
  // Create a Date object to represent today's date
  const today = new Date();
  // Call the fetchAPI function with today's date to fetch available times
  const availableTimes = fakeAPI.fetchAPI(today);
  // Return the fetched available times
  return availableTimes;
};

const updateTimes = (state, selectedDate) => {
  // Call the fetchAPI function with the selected date to fetch available times
  const availableTimes = fakeAPI.fetchAPI(selectedDate);

  // Return the fetched available times
  return availableTimes;
};

function App() {
  // Use reducer to manage state changes for availableTimes
  const [availableTimes, dispatch] = useReducer(
    updateTimesReducer,
    [],
    initializeTimes
  );

  const [date, setDate] = useState("");
  const [time, setTime] = useState("17:00");
  const [guests, setGuests] = useState(1);
  const [occasion, setOccasion] = useState("Birthday");
  const navigate = useNavigate(); // Initialize useNavigate hook
  const [name, setName] = useState("");

  // useEffect to initialize availableTimes when component mounts
  useEffect(() => {
    dispatch({ type: "INITIALIZE_TIMES" });
  }, []); // Empty dependency array ensures this effect runs only once on mount

  // Function to handle state change based on selected date
  const handleDateChange = (selectedDate) => {
    // Dispatch action to update availableTimes based on selectedDate
    dispatch({ type: "UPDATE_TIMES", payload: selectedDate });
    setDate(selectedDate); // Update the date state
  };

  // Function to submit form data
  const submitForm = (formData) => {
    // Update local storage
    localStorage.setItem("formData", JSON.stringify(formData));
    // Call submitAPI with form data and check if it returns true
    if (fakeAPI.submitAPI(formData)) {
      // If submission is successful, navigate to the confirmed booking page
      navigate("/booking-confirmed");
    }
  };

  return (
    <div className="App">
      {/* Render BookingForm component with props */}
      <Routes>
        <Route path="/" element={<MainPage />} />
        <Route
          path="/booking"
          element={
            <BookingForm
              date={date}
              name={name}
              setName={setName}
              setDate={setDate}
              time={time}
              setTime={setTime}
              guests={guests}
              setGuests={setGuests}
              occasion={occasion}
              setOccasion={setOccasion}
              availableTimes={availableTimes}
              updateTimes={handleDateChange} // Pass handleDateChange as updateTimes
              submitForm={submitForm} // Pass submitForm as onSubmit
            />
          }
        />
        <Route path="/booking-confirmed" element={<ConfirmedBooking />} />{" "}
        {/* Define route for confirmed booking page */}
      </Routes>
    </div>
  );
}

export default App;
