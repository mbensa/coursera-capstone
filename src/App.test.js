import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import App, { initializeTimes } from "./App";
import * as FAPI from "./components/FakeAPI";
import { useEffect } from "react";
import * as RN from "react-router-dom";

jest.mock("./components/BookingForm", () => {
  const React = jest.requireActual("react");

  return ({ submitForm }) => {
    React.useEffect(() => {
      submitForm?.();
    }, []);

    return "MockedBookingForm";
  };
});

describe("<App />", () => {
  it("should render", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should correctly initialize times", () => {
    jest.spyOn(Date, "now").mockReturnValue("2024-01-01");
    jest
      .spyOn(FAPI.default, "fetchAPI")
      .mockReturnValue([
        "--- Select a Time ---",
        "15:00",
        "15:30",
        "16:30",
        "17:30",
        "20:00",
        "21:00",
        "22:00",
        "22:30",
        "23:00",
        "23:30",
      ]);

    const result = initializeTimes();

    expect(result).toEqual([
      "--- Select a Time ---",
      "15:00",
      "15:30",
      "16:30",
      "17:30",
      "20:00",
      "21:00",
      "22:00",
      "22:30",
      "23:00",
      "23:30",
    ]);
  });

  it("should save form data to localStorage", () => {
    const setItemSpy = jest.fn();
    Storage.prototype.setItem = setItemSpy;

    const { getByText } = render(
      <MemoryRouter initialEntries={["/booking"]}>
        <App />
      </MemoryRouter>
    );

    expect(setItemSpy).toHaveBeenCalledTimes(1);

    expect(getByText("Your booking has been confirmed.")).toBeInTheDocument();
  });
});
