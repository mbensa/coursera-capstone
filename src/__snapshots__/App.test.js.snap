// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <header>
        <nav
          class="navbar"
        >
          <div
            class="hamburger"
          >
            <span
              class="bar"
            />
            <span
              class="bar"
            />
            <span
              class="bar"
            />
          </div>
          <svg>
            Logo.svg
          </svg>
          <ul
            class="nav-menu"
          >
            <li
              class="nav-item"
            >
              <a
                href="/"
              >
                <h5>
                  Home
                </h5>
              </a>
            </li>
            <li
              class="nav-item"
            >
              <h5>
                About
              </h5>
            </li>
            <li
              class="nav-item"
            >
              <h5>
                Menu
              </h5>
            </li>
            <li
              class="nav-item"
            >
              <a
                href="/booking"
              >
                <h5>
                  Reservations
                </h5>
              </a>
            </li>
            <li
              class="nav-item"
            >
              <h5>
                Order Online
              </h5>
            </li>
            <li
              class="nav-item"
            >
              <h5>
                Login
              </h5>
            </li>
          </ul>
          <svg
            class="basketLogo"
          >
            basket.svg
          </svg>
        </nav>
        <div
          class="headerBg"
        >
          <div
            class="headerContainer"
          >
            <div
              class="mainContainer"
            >
              <h1>
                Little Lemon
              </h1>
              <h2
                class="textColor"
              >
                Chicago
              </h2>
              <h4
                class="textColor descriptionTextMobile"
              >
                We are a family owned Mediterranean restaurant, focused on traditional recipes with a modern twist.
              </h4>
              <a
                href="/booking"
              >
                <div
                  class="btn"
                >
                  <h5>
                    Reserve a Table
                  </h5>
                </div>
              </a>
            </div>
            <div>
              <img
                alt="a man holding a tray with food"
                class="headerImg"
                src="restaurantFood.jpg"
              />
            </div>
          </div>
        </div>
      </header>
      <main>
        <section
          class="sectionContainer"
        >
          <h2>
            This Weeks Specials
          </h2>
          <div
            class="cardMainContainer"
          >
            <div
              class="cardContainer"
            >
              <div
                class="imgContainer imgOne"
              />
              <div
                class="textContainer"
              >
                <div
                  class="dishTitle"
                >
                  <h5>
                    Greek Salad
                  </h5>
                  <h6
                    class="dishPrice"
                  >
                    $12.99
                  </h6>
                </div>
                <p>
                  The famous greek salad of crispy lettuce, peppers, olives and our Chicago style feta cheese, garnished with crunchy garlic and rosemary croutons.
                </p>
                <div
                  class="deliveryContainer"
                >
                  <h6
                    class="deliveryText"
                  >
                    Order a delivery
                  </h6>
                  <img
                    alt="delivery motorbike"
                    src="delivery.jpg"
                  />
                </div>
              </div>
            </div>
            <div
              class="cardContainer"
            >
              <div
                class="imgContainer imgTwo"
              />
              <div
                class="textContainer"
              >
                <div
                  class="dishTitle"
                >
                  <h5>
                    Bruschetta
                  </h5>
                  <h6
                    class="dishPrice"
                  >
                    $7.99
                  </h6>
                </div>
                <p>
                  Our Bruschetta is made from grilled bread that has been smeared with garlic and seasoned with salt and olive oil.
                </p>
                <div
                  class="deliveryContainer"
                >
                  <h6
                    class="deliveryText"
                  >
                    Order a delivery
                  </h6>
                  <img
                    alt="delivery motorbike"
                    src="delivery.jpg"
                  />
                </div>
              </div>
            </div>
            <div
              class="cardContainer"
            >
              <div
                class="imgContainer imgThree"
              />
              <div
                class="textContainer"
              >
                <div
                  class="dishTitle"
                >
                  <h5>
                    Lemon Dessert
                  </h5>
                  <h6
                    class="dishPrice"
                  >
                    $5.00
                  </h6>
                </div>
                <p>
                  This comes straight from grandmaâ€™s recipe book, every last ingredient has been sourced and is as authentic as can be imagined.
                </p>
                <div
                  class="deliveryContainer"
                >
                  <h6
                    class="deliveryText"
                  >
                    Order a delivery
                  </h6>
                  <img
                    alt="delivery motorbike"
                    src="delivery.jpg"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="btn"
          >
            <h5>
              Online Menu
            </h5>
          </div>
        </section>
        <section
          class="sectionTestimonialsContainer"
        >
          <h2
            class="testimonialsTitle"
          >
            What customers say about us
          </h2>
          <div
            class="testimonialsContainer"
          >
            <div
              class="testimonialscardContainer"
            >
              <h6>
                Rating: 5
              </h6>
              <div
                class="testimonialsImgContainer"
              >
                <img
                  alt="a young woman with long black hair"
                  class="testimonialsImg"
                  src="persona1.jpg"
                />
                <h6>
                  Jamie
                </h6>
              </div>
              <p>
                "Best dining experience I've ever had. Highly recommend."
              </p>
            </div>
            <div
              class="testimonialscardContainer"
            >
              <h6>
                Rating: 5
              </h6>
              <div
                class="testimonialsImgContainer"
              >
                <img
                  alt="a man with glasses and short brown hair"
                  class="testimonialsImg"
                  src="persona2.jpg"
                />
                <h6>
                  Tom
                </h6>
              </div>
              <p>
                "The bruschetta was out of this world. I will definitely return"
              </p>
            </div>
            <div
              class="testimonialscardContainer"
            >
              <h6>
                Rating: 5
              </h6>
              <div
                class="testimonialsImgContainer"
              >
                <img
                  alt="a woman with long curly black hair and glasses"
                  class="testimonialsImg"
                  src="persona3.jpg"
                />
                <h6>
                  Katie
                </h6>
              </div>
              <p>
                "Don't skip the dessert, the lemon pie was absolutely amazing."
              </p>
            </div>
            <div
              class="testimonialscardContainer"
            >
              <h6>
                Rating: 5
              </h6>
              <div
                class="testimonialsImgContainer"
              >
                <img
                  alt="a man with short brown hair and a beard"
                  class="testimonialsImg"
                  src="persona4.jpg"
                />
                <h6>
                  Rick
                </h6>
              </div>
              <p>
                "Thanks for the amazing dining experience. We will gladly return."
              </p>
            </div>
          </div>
        </section>
        <section
          class="aboutContainer"
        >
          <div
            class="aboutTextContainer"
          >
            <h1>
              Little Lemon
            </h1>
            <h2>
              Chicago
            </h2>
            <p>
              Little Lemon is run by two famous chefs, Mario and Adrian. The restaurant has opened its doors in 2003. We serve Mediterranean food with a modern twist. 
            </p>
          </div>
          <div
            class="aboutImgContainer"
          />
        </section>
      </main>
      <footer
        class="footerContainer"
      >
        <svg>
          Logo.svg
        </svg>
        <div
          class="socialContainer"
        >
          <svg
            class="facebook social"
          >
            facebook.svg
          </svg>
          <svg
            class="youtube social"
          >
            youtube.svg
          </svg>
          <svg
            class="twitter social"
          >
            twitter.svg
          </svg>
          <svg
            class="instagram social"
          >
            instagram.svg
          </svg>
        </div>
      </footer>
    </div>
  </div>
</DocumentFragment>
`;
